File: frontend/lib/models/requests/register_request.dart
// ignore_for_file: invalid_annotation_target

import 'package:freezed_annotation/freezed_annotation.dart';

part 'register_request.freezed.dart';
part 'register_request.g.dart';

@freezed
class RegisterRequest with _$RegisterRequest {
  factory RegisterRequest({
    @JsonKey(name: 'first_name') required String firstName,
    @JsonKey(name: 'last_name') required String lastName,
    @JsonKey(name: 'phone_number') required String phoneNumber,
    @JsonKey(name: 'password') required String password,
    @JsonKey(name: "password_confirmation")
    required String passwordConfirmation,
  }) = _RegisterRequest;

  factory RegisterRequest.fromJson(Map<String, dynamic> json) =>
      _$RegisterRequestFromJson(json);
}

File: backend/app/Http/Requests/AuthRequests/RegisterRequest.php
<?php

namespace App\Http\Requests\AuthRequests;

use Illuminate\Contracts\Validation\Validator;
use Illuminate\Http\Exceptions\HttpResponseException;

use Illuminate\Foundation\Http\FormRequest;
use Illuminate\Http\JsonResponse;

class RegisterRequest extends FormRequest
{
    /**
     * Determine if the user is authorized to make this request.
     */
    public function authorize(): bool
    {
        return true;
    }

    /**
     * Get the validation rules that apply to the request.
     *
     * @return array<string, \Illuminate\Contracts\Validation\ValidationRule|array<mixed>|string>
     */
    public function rules(): array
    {
        return [
            'first_name'=>'required|string',
            'last_name'=>'required|string',
            'phone_number'=>'required|string|min:10|max:10|unique:users,phone',
            'password' => 'required|string|confirmed|min:8',
        ];
    }

    /**
     * failed validation response
     * @param Validator $validator
     */
    protected function failedValidation(Validator $validator)
    {
        $errors = $validator->errors();

        $customResponse = response()->json([
            'success' => false,
            'message' => 'Invalid Credentials',
            'errors' => $errors,
            'status_message'=>'HTTP_BAD_REQUEST',
        ],JsonResponse::HTTP_BAD_REQUEST);

        throw new HttpResponseException($customResponse);
    }
    /**
     * messages for authentication errors
     */
    public function messages()
    {
        return [
            'first_name.required'=>'Please provide your first name.',
            'last_name.required'=>'Please provide your last name.',
            'phone_number.required' => 'Please provide your phone number.',
            'password.required' => 'Your password is required.',
        ];
    }
}

